Update spray.el to catch runaway timer

TroyFletcher opened the pull request:
Prevent spray--update from spraying in list-timers. If the timer gets stuck it is very difficult to stop without this.

I understand this is a mirror, but the upstream is 404'd. Please advise.

Thanks for your interest in contributing to this package.

This repository is only a mirror.

⚠️ Please do not open a pull-request here! ⚠️ 

🆘 Visit https://emacsmirror.net/stats/upstreams.html to lookup the homepage or upstream repository of this package.

(If you are opening a pull-request at https://github.com/emacsattic/.github, then please disregard the above.  It does not apply in that case, but for technical reasons I cannot prevent it from being shown to you anyway.)

tarsius commented:
> I understand this is a mirror, but the upstream is 404'd.

That must have happened very recently. Like today or yesterday. So hopefully it is okay to ping @ian-kelling still.

> Please advise.

I don't understand this change but it looks wrong to me.

- I cannot find a `timer-list-mode` in my emacs and this package does not depend on any third-party packages that might bring that in as a dependency.
- This is not how one tests if a major-mode is enabled. Use `derived-mode-p`, or even just `eq`.
- This file uses spaces for indentation. Your edit uses tabs.
- The commit message hints at what kind of symptom is supposed to address, but it does not explain what the cause is and/or how your change addresses that.
- Generally, if you want someone to help you with a package they have absolutely not relations with, you should put in a bit more effort. Ideally I would be able to go: "Ah, that obviously is correct. Merge and done."

TroyFletcher commented:
>     * This is not how one tests if a major-mode is enabled. Use `derived-mode-p`, or even just `eq`.

Using derived-mode-p is fine, assuming this fix even makes sense after the issue section below.

I have no excuse for failing to use `eq` my apologies.

>     * This file uses spaces for indentation. Your edit uses tabs.

Apologies, I failed to notice.

>     * The commit message hints at what kind of symptom is supposed to address, but it does not explain what the cause is and/or how your change addresses that.

# Issue
In the years I have been using spray-mode, I have managed to encounter a runaway spray--update timer no less than 3 times on different emacs versions back to 26.3, and most recently on 28.2

I am unaware of how I cause it, but suspect it has to do with an interaction with eww, evil, and/or helm.

The symptoms can be emulated by evaluating `(run-with-timer 0 (/ 60.0 spray-wpm) 'spray--update)`

## Symptoms

* The timer aggressively narrows the margin in all active buffers until `(eobt)`
* spray--update complains of insufficient arguments in all input buffers including the mini-buffer and `(y-or-n-p)` but process the input intermittently.
* Evaluating `(spray-quit)` `(spray-start/stop)` `(spray-stop)` `(cancel-timer spray--running)` has no effect.
* Activating and deactivating spray-mode affects the base-face, but the timer remains active.
* Decreasing variable `spray-wpm` is ignored

# Workaround

Remove the timer with the disabled command `(list-timers)` disable read only mode for the buffer and delete lines until `(eobt)` is true on the line with the (spray--update) timer, then press `c` or evaluate `(timer-list-cancel)`

# Root cause

Some unknown means of exiting spray-mode without `(cancel-timer spray--running)` or otherwise removing the timer indicates spray mode needs to track exit states more robustly.

# Fix

Without knowing the root cause, the symptom can be treated by preventing spray--update from running on buffers named `*timer-list*`, or buffers where `timer-list-mode` is the major mode.

Checking the major mode is an unobtrusive way to check the current buffer with no additional dependencies.

>     * Generally, if you want someone to help you with a package they have absolutely not relations with, you should put in a bit more effort. Ideally I would be able to go: "Ah, that obviously is correct. Merge and done."

Once again, my apologies. Spray--update already had conditions for not running, and this was a narrow and pointed update that would only impact a disabled command where spray-mode doesn't even work under normal circumstances.

How would you like me to proceed, given the above?

tarsius commented:
`spray-start` needs to record in what buffer it is running, and `spray-update` needs to make that buffer current before doing anything else (and if it has disappeared for some reason, then it must stop the timer).

TroyFletcher commented:
Corrected indentation, buffer detection, and changed fail condition to `(spray-quit)`.

Switching buffers in the middle of a spray now has no impact on the new buffer, and freezes the previous spray which may be recovered with `(spray-start/stop)` or `(spray-quit)`.

Switching between multiple spray buffers behaves as expected.

tarsius commented:
> I understand this is a mirror, but the upstream is 404'd. Please advise.

The upstream, https://git.sr.ht/~iank/spray, is up again. Please contact them. By email, if you cannot find any other contact information.
